azure-iot-device/azure/iot/device/iothub/pipeline/pipeline_stages_method_invoke_http.py

```diff
@@ -0,0 +1,131 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import logging
+import json
+import six.moves.urllib as urllib
+from azure.iot.device.common.pipeline import (
+    pipeline_events_base,
+    pipeline_ops_base,
+    pipeline_ops_http,
+    PipelineStage,
+    pipeline_thread,
+)
+from . import pipeline_ops_iothub, pipeline_ops_method_invoke
+from . import constant as pipeline_constant
+from . import exceptions as pipeline_exceptions
+from azure.iot.device import constant as pkg_constant
+
+logger = logging.getLogger(__name__)
+
+
+class MethodInvokeHTTPTranslationStage(PipelineStage):
+    """
+    PipelineStage which converts other Iot and EdgeHub operations into HTTP operations.  This stage also
+    converts http pipeline events into Iot and EdgeHub pipeline events.
+    """
+
+    def __init__(self):
+        super(MethodInvokeHTTPTranslationStage, self).__init__()
+        self.feature_to_topic = {}
+        self.device_id = None
+        self.module_id = None
+        self.hostname = None
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _execute_op(self, op):
+        def map_http_error(error, http_op):
+            if error:
+                return error
+            elif http_op.status_code >= 300:
+                # TODO map error codes to correct exceptions
+                logger.error("Error {} received from request operation".format(http_op.status_code))
+                logger.error("response body: {}".format(http_op.response_body))
+                return pipeline_exceptions.PipelineError(
+                    "http operation returned status {}".format(http_op.status_code)
+                )
+
+        if isinstance(op, pipeline_ops_iothub.SetIoTHubConnectionArgsOperation):
+            self.device_id = op.device_id
+            self.module_id = op.module_id
+            if op.module_id:
+                client_id = "{}/{}".format(op.device_id, op.module_id)
+            else:
+                client_id = op.device_id
```

Comment:
client_id is an MQTT concept, and I don't think there is a specific HTTP analog.  It's possible that we don't need an SetHTTPConnectionArgsOperation object and we can just let the HTTPTransportStage use the SetIoTHubConnectionArgsOperation object directly.

Reply:
I've removed client_id but I'm keeping SetIoTHubConnectionArgsOperation separate for now. It has enough distinctions that I'm more comfortable with that.

---
In reply to: [352772765](https://github.com/Azure/azure-iot-sdk-python/pull/403#discussion_r352772765) [](ancestors = 352772765)